### PR TITLE
Add user metadata

### PR DESCRIPTION
**Pull request by maeldebon**
🔗 **Related PRs:**
- [substra-tests](https://github.com/SubstraFoundation/substra-tests/pull/90)
- [substra](https://github.com/SubstraFoundation/substra/pull/143)
- [substra-backend](https://github.com/SubstraFoundation/substra-backend/pull/186)
- [substra-frontend (coming soon)](#)

The goal of these PRs is to add a `metadata` field in every asset in order to be able to store information related to this asset. It could be, for instance, a creation_ts, an epoch, a project_name, ...

### CODE REVIEW

File: chaincode/algo_aggregate_test.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package main
 
 import (
 	"encoding/json"
+	"fmt"
```
Comment (maeldebon) [inline chaincode/algo_aggregate_test.go]: Oops, shouldn't be here, I'll remove this in the next commit

File: chaincode/input.go [hunk 2]
```diff
@@ -31,42 +31,46 @@ var (
...
-	Permissions               inputPermissions `validate:"required" json:"permissions"`
+	Name                      string           	`validate:"required,gte=1,lte=100" json:"name"`
+	DescriptionHash           string           	`validate:"required,len=64,hexadecimal" json:"descriptionHash"`
+	DescriptionStorageAddress string           	`validate:"required,url" json:"descriptionStorageAddress"`
+	MetricsName               string           	`validate:"required,gte=1,lte=100" json:"metricsName"`
+	MetricsHash               string           	`validate:"required,len=64,hexadecimal" json:"metricsHash"`
+	MetricsStorageAddress     string           	`validate:"required,url" json:"metricsStorageAddress"`
+	TestDataset               inputDataset     	`validate:"omitempty" json:"testDataset"`
+	Permissions               inputPermissions 	`validate:"required" json:"permissions"`
+	Metadata				  map[string]string	`validate:"omitempty" json:"metadata"`
```
Comment (thibaultrobert) [inline chaincode/input.go]: It's strange that your editor didn't take care of it but usually the different section of a struct validation are aligned.

Comment (thibaultrobert) [reply chaincode/input.go]: Got the problem: you've used tabs between names, types, and tags. It should have been spaces.

Comment (thibaultrobert) [reply chaincode/input.go]: I formatted all go file, see the diff: https://gist.github.com/thibaultrobert/1e142684add2418739b6793c0a3103b6
You can apply it or try to tweak the formatting tools of your editor/IDE.

Comment (maeldebon) [reply chaincode/input.go]: Yes, thank you! It should be better like this

**Comment by samlesu** [comment]
Could we limit the number of items in the dict and the length of the keys / values?

**Comment by jmorel** [comment]
I added a Dataset without any metadata and got in return `"metadata": null`. Would it be possible to always return a dict instead, and make it empty if no input data was sent?

File: chaincode/input.go [hunk 3]
```diff
@@ -31,42 +31,46 @@ var (
...
+	Metadata                  map[string]string `validate:"omitempty,lte=100,dive,keys,lte=50,endkeys,lte=100" json:"metadata"`
 }
 
 // inputDataset is the representation in input args to register a dataset
 type inputDataset struct {
-	DataManagerKey string   `validate:"omitempty,len=64,hexadecimal" json:"dataManagerKey"`
-	DataSampleKeys []string `validate:"omitempty,dive,len=64,hexadecimal" json:"dataSampleKeys"`
+	DataManagerKey string             `validate:"omitempty,len=64,hexadecimal" json:"dataManagerKey"`
+	DataSampleKeys []string           `validate:"omitempty,dive,len=64,hexadecimal" json:"dataSampleKeys"`
+	Metadata       map[string]string  `validate:"omitempty,lte=100,dive,keys,lte=50,endkeys,lte=100" json:"metadata"`
```
Comment (samlesu) [inline chaincode/input.go]: As far as I know, inputDataset represents an internal object and not a dataset asset. In the chaincode, the dataset asset is identified as `dataManager`.
@thibaultrobert could you confirm?

**Review by samlesu — DISMISSED** [review]

File: chaincode/input.go (hunk 3 above)
Comment (thibaultrobert) [reply chaincode/input.go]: Yes you right it's only use as a part of an `inputObjective`. It represent the combination of one `dataManager` and several `dataSample`.
Adding a Metadata field seems a bit overkill.

**Review by thibaultrobert — CHANGES_REQUESTED** [review]
I think `initMapOutput` should be used to make sure the `Metadata` fields is set in the output only, not in the ledger itself.

File: EXAMPLES.md [hunk 4]
```diff
@@ -292,6 +298,7 @@ peer chaincode query -n mycc -c '{"Args":["queryObjectives"]}' -C myc
     "bb1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "bb2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: You've still have some null metadata in output

File: EXAMPLES.md [hunk 5]
```diff
@@ -383,12 +392,14 @@ peer chaincode invoke -n mycc -c '{"Args":["queryFilter","{\"indexName\":\"train
     "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 6]
```diff
@@ -481,12 +494,14 @@ peer chaincode invoke -n mycc -c '{"Args":["logSuccessTrain","{\"key\":\"ebbf6cd
    "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
    "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
   ],
+  "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 7]
```diff
@@ -430,12 +441,14 @@ peer chaincode invoke -n mycc -c '{"Args":["logStartTrain","{\"key\":\"ebbf6cdde
...
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 8]
```diff
@@ -530,12 +545,14 @@ peer chaincode invoke -n mycc -c '{"Args":["queryTraintuple","{\"key\":\"ebbf6cd
...
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 9]
```diff
@@ -1060,12 +1090,14 @@ peer chaincode query -n mycc -c '{"Args":["queryModelDetails","{\"key\":\"ebbf6c
     "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 10]
```diff
@@ -1104,12 +1136,14 @@ peer chaincode query -n mycc -c '{"Args":["queryModels"]}' -C myc
      "aa1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
      "aa2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
     ],
+    "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 11]
```diff
@@ -1139,6 +1173,7 @@ peer chaincode query -n mycc -c '{"Args":["queryModels"]}' -C myc
...
```
Comment (thibaultrobert) [inline EXAMPLES.md]: Here too.

File: EXAMPLES.md [hunk 12]
```diff
@@ -1497,6 +1536,7 @@ peer chaincode invoke -n mycc -c '{"Args":["queryObjectiveLeaderboard","{\"objec
     "bb1bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc",
     "bb2bb7c31f62244c0f3a761cc168804227115793d01c270021fe3f7935482dcc"
    ],
+   "metadata": null,
```
Comment (thibaultrobert) [inline EXAMPLES.md]: And here.

**Review by thibaultrobert — CHANGES_REQUESTED** [review]
Sorry as stated in my comments you've still have some metadata that are at `null` in the output.  It seems due to embedded structures so it might not be as critical as the others but it's still incoherent.

**Review by thibaultrobert — APPROVED** [review]

**Review by thibaultrobert — APPROVED** [review]

**Pull request merged**